akka-http-core/src/main/scala/akka/http/impl/model/parser/SimpleHeaders.scala

```diff
@@ -194,8 +194,21 @@ private[parser] trait SimpleHeaders { this: Parser with CommonRules with CommonA
   // http://tools.ietf.org/html/rfc7231#section-7.4.2
   def server = rule { products ~ EOI ~> (Server(_)) }
 
-  def `strict-transport-security` = rule {
-    ignoreCase("max-age=") ~ `delta-seconds` ~ optional(ws(";") ~ ignoreCase("includesubdomains") ~ push(true)) ~ zeroOrMore(ws(";") ~ token0) ~ EOI ~> (`Strict-Transport-Security`(_, _))
+  // https://tools.ietf.org/html/rfc6797#section-6
+  def `strict-transport-security` = {
+    def directives =
+      rule(
+        ignoreCase("includesubdomains") ~ push(IncludeSubDomains)
+          | ignoreCase("max-age=") ~ `delta-seconds` ~> (MaxAge(_))
+      )
+    def ignoredDirective = rule {
+      token ~ optional(ws("=") ~ word) ~> ((k: String, v: Option[String]) ⇒ IgnoredDirective(k + v.getOrElse("")))
+    }
+
+    rule {
+      oneOrMore(directives | ignoredDirective).separatedBy(oneOrMore(ws(";"))) ~ zeroOrMore(ws(";")) ~ EOI ~>
+        (`Strict-Transport-Security`.fromDirectives(_: _*))
+    }
```

Comment:
In the spec it says `[ directive ]  *( ";" [ directive ] )` unless I'm really tired this morning and am missing something, that doesn't mean that a trailing semicolon is allowed. It means a directive followed by zero or more semicolon prefixed directives.

Also, you didn't mention adding accepting ignoring unknown directives, is this really a good idea? I see only a note about non-compliant UAs ignoring future/unknown directives in the spec.

Reply:
Hi, thanks for your comments!

Going by the ABNF spec, `[ directive ] *( ";" [ directive ] )` can be rewritten to `0*1( directive )  0*∞( ";" 0*1( directive ) )`. Especially note that the directives are *optional*, and therefore in my understanding the following strings are all syntactically valid:
* directive1; directive2
* directive1;; directive2;
* ; directive2
* ;;;
However, they are not conform spec, because the spec requires at least the `max-age` directive to be present. Nevertheless, I think a trailing semicolon is syntactically allowed!

I'm not sure I understand your point about ignored directives. [Section 6.1 of the spec](https://tools.ietf.org/html/rfc6797#section-6.1) mentions:

> 5.  If an STS header field contains directive(s) not recognized by
>     the UA, the UA MUST ignore the unrecognized directives, and if
>     the STS header field otherwise satisfies the above requirements
>     (1 through 4), the UA MUST process the recognized directives.

In my implementation, ignored directives are parsed into instances of `IgnoredDirective`, and filtered out in `Strict-Transport-Security#fromDirectives`. Does this clarify it for you?

Reply:
Yes, you are right, the square brackets is what my monday morning eyes didn't catch, thanks for clarifying.

Reply:
And point 5. Good, then this is the expected behavior per spec!

Reply:
(really weird spec though... but yeah)

Reply:
I'm not entirely sure about those `IgnoredDirective`s: indeed the spec says the UA must ignore unrecognized directives, but even if Akka HTTP doesn't recognize them, the application could, so I'm not sure it's the responsibility of Akka HTTP to filter those out. OTOH we don't keep arbitrary fields in e.g. `HttpCookie` so this is consistent & fine.